grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java

```diff
@@ -444,41 +449,92 @@ private static String getResponseBody(List<FieldDescriptor> topLevelFields,
     }
 
     @Nullable
-    private static Function<HttpData, HttpData> generateResponseBodyConverter(TranscodingSpec spec) {
-        @Nullable final String responseBody = spec.responseBody;
+    private static Function<AggregatedHttpResponse, AggregatedHttpResponse> generateResponseConverter(
+            TranscodingSpec spec) {
+        // Ignore the spec if the method is HttpBody. The response body is already in the correct format.
+        if (HttpBody.getDescriptor().equals(spec.methodDescriptor.getOutputType())) {
+            return httpResponse -> {
+                final JsonNode jsonNode = extractHttpBody(httpResponse.content());
+
+                // Failed to parse the JSON body, return the original response.
+                if (jsonNode == null) {
+                    return httpResponse;
+                }
+
```

Comment:
Can we also close the `httpResponse.content()` once we're sure we will be using `google.api.httpBody`? (Either via `try` blocks, or by using `PooledObjects.close(httpResponse.content())`)

Reply:
I guess I got it, I am not entirely sure how that `try(...)` works internally or what pooled objects are TBH. Please LMK if it looks right.

Reply:
> I am not entirely sure how that try(...) works internally or what pooled objects are TBH.

Once a gRPC message is written, it is converted to bytes via `MessageMarshaller`

https://github.com/line/armeria/blob/ebb08880ab8aa87f1d6ffe8e0dcf31d09b11544b/grpc/src/main/java/com/linecorp/armeria/internal/server/grpc/AbstractServerCall.java#L547

Here the channel's `ByteBufAllocator` is used to allocate bytes so that no additional copies are necessary when writing over the wire.

https://github.com/line/armeria/blob/ebb08880ab8aa87f1d6ffe8e0dcf31d09b11544b/grpc/src/main/java/com/linecorp/armeria/internal/common/grpc/GrpcMessageMarshaller.java#L145

The response is aggregated with the `pooled` option, meaning don't copy the bytes
https://github.com/line/armeria/blob/ebb08880ab8aa87f1d6ffe8e0dcf31d09b11544b/grpc/src/main/java/com/linecorp/armeria/server/grpc/AbstractUnframedGrpcService.java#L172

Hence, if we want to use the `httpResponse` as-is in the early return, we shouldn't close the `httpResponse.content()`.

Can you remove the try block here and modify like the following?

e.g.
```java
return httpResponse -> {
    final HttpData data = httpResponse.content();
    final JsonNode jsonNode = extractHttpBody(data);

    // Failed to parse the JSON body, return the original response.
    if (jsonNode == null) {
        return httpResponse;
    }
    PooledObjects.close(data);

    // The data field is base64 encoded.
    // https://protobuf.dev/programming-guides/proto3/#json
```

Reply:
Okay, I guess I kinda understand. What happens if we don't close the pool? I suspect the memory has to be claimed by the garbage collector which is inefficient? Or maybe there is some reference to this byte buffer left somewhere internally thus it might lead to a memory leak? (Seems unlikely)

Asking just for learning purposes 👼

Reply:
This is more netty territory, but my understanding is that since direct memory is not cheap to allocate/deallocate netty maintains a pool of java `ByteBuffer`s (may be direct or heap based)

If we don't call `release` on the `ByteBuf`, the `ByteBuf` object itself will be garbage collected but the memory pool will continue to have a strong reference to the underlying`ByteBuffer` - hence a memory leak. The others may shed more light on this subject though.